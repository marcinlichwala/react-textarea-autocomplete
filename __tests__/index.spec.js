import React, { Component } from 'react';
import { shallow, mount } from 'enzyme';
import ReactTextareaAutocomplete from '../src';

//eslint-disable-next-line
const SmileItemComponent = ({ entity: { label, text } }) => (
  <div> {label} </div>
);

const Loading = () => <div>Loading...</div>;

describe('object-based items', () => {
  const mockedFn = jest.fn();
  const rtaComponent = (
    <ReactTextareaAutocomplete
      placeholder={'Write a message.'}
      value={'Controlled text'}
      onChange={mockedFn}
      className={'ownClassName'}
      style={{ background: 'red' }}
      loadingComponent={Loading}
      trigger={{
        ':': {
          output: item => `___${item.text}___`,
          dataProvider: () => [
            { id: 1, label: ':)', text: 'happy_face' },
            { id: 2, label: ':(', text: 'sad_face' },
          ],
          component: SmileItemComponent,
        },
      }}
    />
  );
  const rta = mount(rtaComponent);

  it('match match the snapshot', () => {
    expect(shallow(rtaComponent)).toMatchSnapshot();
  });

  it('Textarea exists', () => {
    expect(rta.find('textarea')).toHaveLength(1);
  });

  it('After the trigger was typed, it should appear list of options', async () => {
    rta
      .find('textarea')
      .simulate('change', { target: { value: 'some test :a' } });
    expect(rta.find('.rta__autocomplete')).toHaveLength(1);
  });

  it('should display all items', () => {
    expect(rta.find('.rta__autocomplete .rta__list .rta__item')).toHaveLength(
      2,
    );
  });

  it('items should be rendered within the component', () => {
    expect(
      rta
        .find('.rta__item')
        .first()
        .containsMatchingElement(<SmileItemComponent />),
    ).toEqual(true);
  });

  it('should invoke onchange handler', () => {
    expect(mockedFn).toHaveBeenCalled();
  });

  it('should close the autocomplete after mouse click', () => {
    const item = rta.find('.rta__entity').first();
    item.simulate('click');
    expect(rta.find('.rta__entity')).toHaveLength(0);
  });

  it('should invoke onchange handler after selection', () => {
    expect(mockedFn).toHaveBeenCalledTimes(2);
  });

  it('text in textarea should be changed', () => {
    expect(rta.find('textarea').node.value).toBe(
      '___happy_face___some test :a',
    );
  });
});

describe('string-based items w/o output fn', () => {
  const mockedFn = jest.fn();
  const rtaComponent = (
    <ReactTextareaAutocomplete
      placeholder={'Write a message.'}
      value={'Controlled text'}
      onChange={mockedFn}
      className={'ownClassName'}
      style={{ background: 'red' }}
      loadingComponent={Loading}
      trigger={{
        ':': {
          dataProvider: () => Promise.resolve(['happy_face', 'sad_face']),
          component: SmileItemComponent,
        },
      }}
    />
  );
  const rta = mount(rtaComponent);

  it('match match the snapshot', () => {
    expect(shallow(rtaComponent)).toMatchSnapshot();
  });

  it('Textarea exists', () => {
    expect(rta.find('textarea')).toHaveLength(1);
  });

  it('After the trigger was typed, it should appear list of options', async () => {
    rta
      .find('textarea')
      .simulate('change', { target: { value: 'some test :a' } });
    expect(rta.find('.rta__autocomplete')).toHaveLength(1);
  });

  it('should display all items', () => {
    expect(rta.find('.rta__autocomplete .rta__list .rta__item')).toHaveLength(
      2,
    );
  });

  it('items should be rendered within the component', () => {
    expect(
      rta
        .find('.rta__item')
        .first()
        .containsMatchingElement(<SmileItemComponent />),
    ).toEqual(true);
  });

  it('should invoke onchange handler', () => {
    expect(mockedFn).toHaveBeenCalled();
  });

  it('should close the autocomplete after mouse click', () => {
    const item = rta.find('.rta__entity').first();
    item.simulate('click');
    expect(rta.find('.rta__entity')).toHaveLength(0);
  });

  it('should invoke onchange handler after selection', () => {
    expect(mockedFn).toHaveBeenCalledTimes(2);
  });

  it('text in textarea should be changed', () => {
    expect(rta.find('textarea').node.value).toBe(':happy_face:some test :a');
  });
});

describe('string-based items with output fn', () => {
  const mockedFn = jest.fn();
  const rtaComponent = (
    <ReactTextareaAutocomplete
      placeholder={'Write a message.'}
      value={'Controlled text'}
      onChange={mockedFn}
      className={'ownClassName'}
      style={{ background: 'red' }}
      loadingComponent={Loading}
      trigger={{
        ':': {
          output: item => `__${item}__`,
          dataProvider: () => Promise.resolve(['happy_face', 'sad_face']),
          component: SmileItemComponent,
        },
      }}
    />
  );
  const rta = mount(rtaComponent);

  it('match match the snapshot', () => {
    expect(shallow(rtaComponent)).toMatchSnapshot();
  });

  it('Textarea exists', () => {
    expect(rta.find('textarea')).toHaveLength(1);
  });

  it('After the trigger was typed, it should appear list of options', async () => {
    rta
      .find('textarea')
      .simulate('change', { target: { value: 'some test :a' } });
    expect(rta.find('.rta__autocomplete')).toHaveLength(1);
  });

  it('should display all items', () => {
    expect(rta.find('.rta__autocomplete .rta__list .rta__item')).toHaveLength(
      2,
    );
  });

  it('items should be rendered within the component', () => {
    expect(
      rta
        .find('.rta__item')
        .first()
        .containsMatchingElement(<SmileItemComponent />),
    ).toEqual(true);
  });

  it('should invoke onchange handler', () => {
    expect(mockedFn).toHaveBeenCalled();
  });

  it('should close the autocomplete after mouse click', () => {
    const item = rta.find('.rta__entity').first();
    item.simulate('click');
    expect(rta.find('.rta__entity')).toHaveLength(0);
  });

  it('should invoke onchange handler after selection', () => {
    expect(mockedFn).toHaveBeenCalledTimes(2);
  });

  it('text in textarea should be changed', () => {
    expect(rta.find('textarea').node.value).toBe('__happy_face__some test :a');
  });
});

describe('using ref to the ReactTextareaAutocomplete to call methods', () => {
  const mockedFn = jest.fn();

  class ReactTextareaAutocompleteWrapper extends Component {

    getCaretPosition() {
      return this.textareaRef.getCaretPosition();
    }

    setCaretPosition(position) {
      this.textareaRef.setCaretPosition(position);
    }

    render() {
      return (
        <ReactTextareaAutocomplete
          placeholder={'Write a message.'}
          value={'Controlled text'}
          onChange={mockedFn}
          className={'ownClassName'}
          style={{ background: 'red' }}
          loadingComponent={Loading}
          trigger={{
            ':': {
              dataProvider: () => Promise.resolve(['happy_face', 'sad_face']),
              component: SmileItemComponent,
            },
          }}
          ref={(ref) => { this.textareaRef = ref; }}
        />
      );
    }
  }

  const rtaWrapper = mount(<ReactTextareaAutocompleteWrapper />);
  const rtaWrapperRef = rtaWrapper.instance();

  it('should get the correct caret position initially', () => {
    const actual = rtaWrapperRef.getCaretPosition();
    const expected = 0;

    expect(actual).toBe(expected);
  });

  it('should get the correct caret position after typing in some data', () => {
    const someData = 'some data is entered';
    const position = someData.length;

    rtaWrapper.find('textarea')
        .simulate('change', { target: { value: someData } });

    rtaWrapper
        .find('textarea')
        .node.focus();

    const actual = rtaWrapperRef.getCaretPosition();
    const expected = position;

    expect(actual).toBe(expected);
  });

  it('should set the caret position correctly', () => {
    const CARET_POSITION_TO_SET = 5;

    rtaWrapperRef.setCaretPosition(CARET_POSITION_TO_SET);

    const actual = rtaWrapperRef.getCaretPosition();
    const expected = CARET_POSITION_TO_SET;

    expect(actual).toBe(expected);
  });
});
