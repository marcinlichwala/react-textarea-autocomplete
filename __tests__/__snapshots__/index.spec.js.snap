// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`object-based items match match the snapshot 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="rta "
    style={undefined}
  >
    <textarea
      className="rta__textarea ownClassName"
      onChange={[Function]}
      onSelect={[Function]}
      placeholder="Write a message."
      style={
        Object {
          "background": "red",
        }
      }
      value="Controlled text"
    />
  </div>,
  "nodes": Array [
    <div
      className="rta "
      style={undefined}
    >
      <textarea
        className="rta__textarea ownClassName"
        onChange={[Function]}
        onSelect={[Function]}
        placeholder="Write a message."
        style={
          Object {
            "background": "red",
          }
        }
        value="Controlled text"
      />
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <ReactTextareaAutocomplete
        className="ownClassName"
        containerStyle={undefined}
        loadingComponent={[Function]}
        minChar={1}
        onCaretPositionChange={[Function]}
        onChange={[Function]}
        onSelect={[Function]}
        placeholder="Write a message."
        style={
          Object {
            "background": "red",
          }
        }
        trigger={
          Object {
            ":": Object {
              "component": [Function],
              "dataProvider": [Function],
              "output": [Function],
            },
          }
        }
        value="Controlled text"
      />,
      "_debugID": 17,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": ReactTextareaAutocomplete {
        "_changeHandler": [Function],
        "_cleanUpProps": [Function],
        "_closeAutocomplete": [Function],
        "_getCurrentTriggerSettings": [Function],
        "_getSuggestions": [Function],
        "_getTextToReplace": [Function],
        "_getValuesFromProvider": [Function],
        "_onSelect": [Function],
        "_reactInternalInstance": [Circular],
        "_selectHandler": [Function],
        "context": Object {},
        "getCaretPosition": [Function],
        "props": Object {
          "className": "ownClassName",
          "containerStyle": undefined,
          "loadingComponent": [Function],
          "minChar": 1,
          "onCaretPositionChange": [Function],
          "onChange": [Function],
          "onSelect": [Function],
          "placeholder": "Write a message.",
          "style": Object {
            "background": "red",
          },
          "trigger": Object {
            ":": Object {
              "component": [Function],
              "dataProvider": [Function],
              "output": [Function],
            },
          },
          "value": "Controlled text",
        },
        "refs": Object {},
        "setCaretPosition": [Function],
        "state": Object {
          "actualToken": "",
          "component": null,
          "currentTrigger": null,
          "data": null,
          "dataLoading": false,
          "left": 0,
          "selectionEnd": 0,
          "selectionStart": 0,
          "top": 0,
          "value": "Controlled text",
        },
        "tokenRegExp": /\\[:\\]\\\\w\\*\\$/,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="rta "
          style={undefined}
        >
          <textarea
            className="rta__textarea ownClassName"
            onChange={[Function]}
            onSelect={[Function]}
            placeholder="Write a message."
            style={
              Object {
                "background": "red",
              }
            }
            value="Controlled text"
          />
        </div>,
        "_debugID": 18,
        "_renderedOutput": <div
          className="rta "
          style={undefined}
        >
          <textarea
            className="rta__textarea ownClassName"
            onChange={[Function]}
            onSelect={[Function]}
            placeholder="Write a message."
            style={
              Object {
                "background": "red",
              }
            }
            value="Controlled text"
          />
        </div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ReactTextareaAutocomplete
    className="ownClassName"
    containerStyle={undefined}
    loadingComponent={[Function]}
    minChar={1}
    onCaretPositionChange={[Function]}
    onChange={[Function]}
    onSelect={[Function]}
    placeholder="Write a message."
    style={
      Object {
        "background": "red",
      }
    }
    trigger={
      Object {
        ":": Object {
          "component": [Function],
          "dataProvider": [Function],
          "output": [Function],
        },
      }
    }
    value="Controlled text"
  />,
}
`;

exports[`string-based items w/o output fn match match the snapshot 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="rta "
    style={undefined}
  >
    <textarea
      className="rta__textarea ownClassName"
      onChange={[Function]}
      onSelect={[Function]}
      placeholder="Write a message."
      style={
        Object {
          "background": "red",
        }
      }
      value="Controlled text"
    />
  </div>,
  "nodes": Array [
    <div
      className="rta "
      style={undefined}
    >
      <textarea
        className="rta__textarea ownClassName"
        onChange={[Function]}
        onSelect={[Function]}
        placeholder="Write a message."
        style={
          Object {
            "background": "red",
          }
        }
        value="Controlled text"
      />
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <ReactTextareaAutocomplete
        className="ownClassName"
        containerStyle={undefined}
        loadingComponent={[Function]}
        minChar={1}
        onCaretPositionChange={[Function]}
        onChange={[Function]}
        placeholder="Write a message."
        style={
          Object {
            "background": "red",
          }
        }
        trigger={
          Object {
            ":": Object {
              "component": [Function],
              "dataProvider": [Function],
            },
          }
        }
        value="Controlled text"
      />,
      "_debugID": 41,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": ReactTextareaAutocomplete {
        "_changeHandler": [Function],
        "_cleanUpProps": [Function],
        "_closeAutocomplete": [Function],
        "_getCurrentTriggerSettings": [Function],
        "_getSuggestions": [Function],
        "_getTextToReplace": [Function],
        "_getValuesFromProvider": [Function],
        "_onSelect": [Function],
        "_reactInternalInstance": [Circular],
        "_selectHandler": [Function],
        "context": Object {},
        "getCaretPosition": [Function],
        "props": Object {
          "className": "ownClassName",
          "containerStyle": undefined,
          "loadingComponent": [Function],
          "minChar": 1,
          "onCaretPositionChange": [Function],
          "onChange": [Function],
          "placeholder": "Write a message.",
          "style": Object {
            "background": "red",
          },
          "trigger": Object {
            ":": Object {
              "component": [Function],
              "dataProvider": [Function],
            },
          },
          "value": "Controlled text",
        },
        "refs": Object {},
        "setCaretPosition": [Function],
        "state": Object {
          "actualToken": "",
          "component": null,
          "currentTrigger": null,
          "data": null,
          "dataLoading": false,
          "left": 0,
          "selectionEnd": 0,
          "selectionStart": 0,
          "top": 0,
          "value": "Controlled text",
        },
        "tokenRegExp": /\\[:\\]\\\\w\\*\\$/,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="rta "
          style={undefined}
        >
          <textarea
            className="rta__textarea ownClassName"
            onChange={[Function]}
            onSelect={[Function]}
            placeholder="Write a message."
            style={
              Object {
                "background": "red",
              }
            }
            value="Controlled text"
          />
        </div>,
        "_debugID": 42,
        "_renderedOutput": <div
          className="rta "
          style={undefined}
        >
          <textarea
            className="rta__textarea ownClassName"
            onChange={[Function]}
            onSelect={[Function]}
            placeholder="Write a message."
            style={
              Object {
                "background": "red",
              }
            }
            value="Controlled text"
          />
        </div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ReactTextareaAutocomplete
    className="ownClassName"
    containerStyle={undefined}
    loadingComponent={[Function]}
    minChar={1}
    onCaretPositionChange={[Function]}
    onChange={[Function]}
    placeholder="Write a message."
    style={
      Object {
        "background": "red",
      }
    }
    trigger={
      Object {
        ":": Object {
          "component": [Function],
          "dataProvider": [Function],
        },
      }
    }
    value="Controlled text"
  />,
}
`;

exports[`string-based items with output fn match match the snapshot 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="rta "
    style={undefined}
  >
    <textarea
      className="rta__textarea ownClassName"
      onChange={[Function]}
      onSelect={[Function]}
      placeholder="Write a message."
      style={
        Object {
          "background": "red",
        }
      }
      value="Controlled text"
    />
  </div>,
  "nodes": Array [
    <div
      className="rta "
      style={undefined}
    >
      <textarea
        className="rta__textarea ownClassName"
        onChange={[Function]}
        onSelect={[Function]}
        placeholder="Write a message."
        style={
          Object {
            "background": "red",
          }
        }
        value="Controlled text"
      />
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <ReactTextareaAutocomplete
        className="ownClassName"
        containerStyle={undefined}
        loadingComponent={[Function]}
        minChar={1}
        onCaretPositionChange={[Function]}
        onChange={[Function]}
        placeholder="Write a message."
        style={
          Object {
            "background": "red",
          }
        }
        trigger={
          Object {
            ":": Object {
              "component": [Function],
              "dataProvider": [Function],
              "output": [Function],
            },
          }
        }
        value="Controlled text"
      />,
      "_debugID": 63,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": ReactTextareaAutocomplete {
        "_changeHandler": [Function],
        "_cleanUpProps": [Function],
        "_closeAutocomplete": [Function],
        "_getCurrentTriggerSettings": [Function],
        "_getSuggestions": [Function],
        "_getTextToReplace": [Function],
        "_getValuesFromProvider": [Function],
        "_onSelect": [Function],
        "_reactInternalInstance": [Circular],
        "_selectHandler": [Function],
        "context": Object {},
        "getCaretPosition": [Function],
        "props": Object {
          "className": "ownClassName",
          "containerStyle": undefined,
          "loadingComponent": [Function],
          "minChar": 1,
          "onCaretPositionChange": [Function],
          "onChange": [Function],
          "placeholder": "Write a message.",
          "style": Object {
            "background": "red",
          },
          "trigger": Object {
            ":": Object {
              "component": [Function],
              "dataProvider": [Function],
              "output": [Function],
            },
          },
          "value": "Controlled text",
        },
        "refs": Object {},
        "setCaretPosition": [Function],
        "state": Object {
          "actualToken": "",
          "component": null,
          "currentTrigger": null,
          "data": null,
          "dataLoading": false,
          "left": 0,
          "selectionEnd": 0,
          "selectionStart": 0,
          "top": 0,
          "value": "Controlled text",
        },
        "tokenRegExp": /\\[:\\]\\\\w\\*\\$/,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 3,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="rta "
          style={undefined}
        >
          <textarea
            className="rta__textarea ownClassName"
            onChange={[Function]}
            onSelect={[Function]}
            placeholder="Write a message."
            style={
              Object {
                "background": "red",
              }
            }
            value="Controlled text"
          />
        </div>,
        "_debugID": 64,
        "_renderedOutput": <div
          className="rta "
          style={undefined}
        >
          <textarea
            className="rta__textarea ownClassName"
            onChange={[Function]}
            onSelect={[Function]}
            placeholder="Write a message."
            style={
              Object {
                "background": "red",
              }
            }
            value="Controlled text"
          />
        </div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ReactTextareaAutocomplete
    className="ownClassName"
    containerStyle={undefined}
    loadingComponent={[Function]}
    minChar={1}
    onCaretPositionChange={[Function]}
    onChange={[Function]}
    placeholder="Write a message."
    style={
      Object {
        "background": "red",
      }
    }
    trigger={
      Object {
        ":": Object {
          "component": [Function],
          "dataProvider": [Function],
          "output": [Function],
        },
      }
    }
    value="Controlled text"
  />,
}
`;
